scipy/stats/_unuran/unuran_wrapper.pyx.templ

```diff
@@ -0,0 +1,1122 @@
+
+        Examples
+        --------
+        >>> from scipy.stats import TransformedDensityRejection
+        >>> from scipy.stats import norm
+        >>> from math import exp
+        >>> 
+        >>> class MyDist:
+        ...     def pdf(self, x):
+        ...         return exp(-0.5 * x**2)
+        ...     def dpdf(self, x):
+        ...         return -x * exp(-0.5 * x**2)
+        ... 
+        >>> dist = MyDist()
+        >>> rng = TransformedDensityRejection(dist)
+        >>> 
+        >>> rng.ppf_hat(0.5)
+        -0.00018050266342393984
+        >>> norm.ppf(0.5)
+        0.0
+        >>> u = np.linspace(0, 1, num=1000)
+        >>> ppf_hat = rng.ppf_hat(u)
+        """
+        u = np.asarray(u, dtype='d')
+        oshape = u.shape
+        u = u.ravel()
+        # UNU.RAN fills in ends of the support when u < 0 or u > 1 while
+        # SciPy fills in nans. Prefer SciPy behaviour.
+        cond0 = 0 <= u
+        cond1 = u <= 1
+        cond2 = cond0 & cond1
+        goodu = argsreduce(cond2, u)[0]
+        out = np.empty_like(u)
+        cdef double[::1] u_view = np.ascontiguousarray(goodu)
+        cdef double[::1] goodout = np.empty_like(u_view)
+        if cond2.any():
+            self._ppf_hat(&u_view[0], &goodout[0], len(goodu))
+        np.place(out, cond2, goodout)
+        np.place(out, ~cond2, np.nan)
+        return np.asarray(out).reshape(oshape)[()]
+
+
+cdef class DiscreteAliasUrn(Method):
+    r"""
+    Discrete Alias-Urn Method.
+
+    This method is used to sample from univariate discrete distributions with
+    a finite domain. It uses the probability vector of size :math:`N` or a
+    probability mass function with a finite support to generator random
+    numbers from the distribution.
+
+    Parameters
+    ----------
+    pv : array_like, optional
+        Probability vector (PV) of the distribution. If PV isn't available,
+        PMF of the distribution is expected.
+    dist : object, optional
+        An instance of a class with a 'pmf' method. Only required if PV isn't
+        available. The signature of the PMF is expected to be:
+        ``def pmf(k: int, *params) -> float``. i.e it should accecpt a Python
+        integer and return a Python float.
+    params : tuple, optional
+        Parameters that the PMF takes. If a probability vector is available,
+        this parameter is ignored. Default is an empty tuple.
+    domain : int, optional
+        Support of the PMF. If a probability vector is not available, a
+        finite domain must be given. i.e. the PMF must have a finite support.
+    urnfactor : float, optional
+        Size of the urn table *relative* to the size of the probability
+        vector. It must not be less than 1. Larger tables result is faster
+        generation times but require a more expensive setup.
+    seed : int, BitGenerator, Generator, RandomState, SeedSequence, optional
+        Seed for the underlying uniform random number generation.
```

Comment:
Is this needed to initialize the object or should it be a parameter of the `rvs` method (as it is for stats distributions and `NumericalInverseHermite`)? (Same question for `TransformedDensityRejection`, although since I understand less about how that works, I can imagine you might need an RNG when creating the object.)

Reply:
> Is this needed to initialize the object or should it be a parameter of the `rvs` method (as it is for stats distributions and `NumericalInverseHermite`)?

This is needed to initialize UNU.RAN generators. We can remove this from here and just use an unseeded NumPy rng to initialize the UNU.RAN generator. Then, the seed can be set later in the `rvs` method, since, the seed can be changed anytime using seed setter: `rng.seed = 123`:

```python
def rvs(self, size, seed):
    self.seed = seed
    # sample
    ...
```

Reply:
that seems like a good idea to align to stats distributions / NumericalInverseHermite

Reply:
> We can remove this from here and just use an unseeded NumPy rng to initialize the UNU.RAN generator

I'm assuming you suggest using an unseeded RNG because the state of the RNG needed to initialize the UNU.RAN generator doesn't affect the output much. In that case, why _not_ provide a seed (so as to avoid the potential for non-reproducible errors) without exposing control over it? Then the behavior is exactly repeatable for a given seed passed into the `rvs` method?

Reply:
The first question would be if the construction of the UNU.RAN generator draws from its URNG before `rvs()` requests random draws from the target distribution. I'm guessing not. It would be weird if it did. It's just that the UNU.RAN API was designed around having the URNG available when one of these methods is instantiated, so for us to fully instantiate a `Method`, we need an URNG.

But even so, I'd still suggest that people pass it into the constructor once and then just call the `rvs()` method without worrying about passing it in again over and over again. Once during construction is enough.

Reply:
> But even so, I'd still suggest that people pass it into the constructor once and then just call the `rvs()` method without worrying about passing it in again over and over again. Once during construction is enough.

I agree with you here. I guess @mdhaber's point was that if we ever want to add a `qrng` method in the future, we would need a `qmc_engine` parameter (or a similar keyword) otherwise there would an inconsistency between `rvs` and `qrng`. To address this, we could add a `qmc_engine` parameter to the constructor itself, right? In that case, no need to change anything here.

Reply:
Just to point out that the method name in `NumericalInverseHermite` is not `qrng` but `qrvs([size, d, qmc_engine])`

Reply:
There can be an inconsistency between `rvs()` and `qrvs()` in this regard; that's fine. Adding an `rng=` parameter to `rvs()` could be a useful _addition_, but not a replacement for passing one into the constructor.

Reply:
> but not a replacement for passing one into the constructor.

Should an `rng` parameter be added to the initializer of `NumericalInverseHermite` and as a parameter when freezing distributions, then? Not saying that it actually has to be done, but by this reasoning, would that be a desirable thing for which we'd accept an enhancement request?

Reply:
> Should an `rng` parameter be added to the initializer of `NumericalInverseHermite` and as a parameter when freezing distributions, then? Not saying that it actually has to be done, but by this reasoning, would that be a desirable thing for which we'd accept an "enhancement" PR?

I think it would make sense to do that when we shift to the UNU.RAN implementation. Once this is done, I will submit a PR to do that.

Reply:
The implementation now supports the attribute `support` :)

With the updated documentation of the approach if PV is used, is that ok for you (@mdhaber)?